src/useWindowSize.ts

```diff
@@ -0,0 +1,57 @@
+import { useEffect, useState } from 'react';
+
+interface WindowSize {
+  width: number;
+  height: number;
+}
+
+/**
+ * useWindowSize : 브라우저 창의 너비와 높이를 제공하는 훅
+ * @param {boolean} [isDelay=true] 딜레이 여부 판단. Default=true
+ * @param {number} [delayTime=200] 지연 시간(밀리초). isDelay가 true일 때 사용됨.Default=200
+ * @returns {{ width: number, height: number }} 창의 너비와 높이를 담은 객체
+ * @description
+ * 브라우저 창의 사이즈가 변경될 때마다 해당 창의 너비와 높이를 업데이트합니다.
+ * delay유무와 delayTime을 설정할 수 있습니다.
+ * UI의 세밀한 조작 등 동적인 변화에 유용합니다.
+ */
+
+const useWindowSize = (
+  isDelay: boolean = true,
+  delayTime: number = 200
+): WindowSize => {
+  const [windowSize, setWindowSize] = useState<WindowSize>({
+    width: window.innerWidth,
+    height: window.innerHeight,
+  });
+
+  useEffect(() => {
+    let timeoutId: NodeJS.Timeout | undefined;
+    const handleWindowResize = () => {
+      if (timeoutId) clearTimeout(timeoutId);
+
+      const updateWindowSize = () => {
+        setWindowSize({
+          width: window.innerWidth,
+          height: window.innerHeight,
+        });
+      };
+
+      if (isDelay) {
+        timeoutId = setTimeout(updateWindowSize, delayTime);
+      } else {
+        updateWindowSize();
+      }
+    };
+
+    window.addEventListener('resize', handleWindowResize);
+    return () => {
+      window.removeEventListener('resize', handleWindowResize);
+      if (timeoutId) clearTimeout(timeoutId);
+    };
+  }, [isDelay, delayTime]);
```

Comment:
```suggestion
  }, []);
```

[P2] 소현 님, 여기에 의존성으로 주입한 두 개의 props는 한 번만 넘어오는 것으로 파악했는데 제가 파악한 게 맞다면 의존성 배열을 비워놔도 되지 않을까요?

Reply:
수정된 현재 [isClient, delayTime]이고, 

이를 비워두면 `React Hook useEffect has missing dependencies: 'delayTime' and 'isClient'. Either include them or remove the dependency array.eslint[react-hooks/exhaustive-deps](https://github.com/facebook/react/issues/14920)`와 같은 eslint경고가 뜨는 상황이라 
초기값 이후 변경될 일이 없다면 useRef 형식으로 저장하는 것이 더 낫겠네요 수정해보겠습니다

Reply:
또 별개로 궁금한 점이 생겼는데


```ts
  const updateWindowSize = () => {
    setWindowSize({
      width: window.innerWidth,
      height: window.innerHeight,
    });
  };
```
updateWindowsize가 중복되다보니 간단한 로직임에도 함수로 정의했었고, 이제 이걸 좀더 정리하면 handleWindowResize밖으로 분리하는게 더 적절한 방식인거 같아 useCallback을 붙여서 useEffect 밖으로 빼서 사용하는 방법도 있는데

```ts
  const updateWindowSize = useCallback(() => {
    setWindowSize({
      width: window.innerWidth,
      height: window.innerHeight,
    });
  }, []);

  useEffect(() => {
    if (!clientRef.current) {
      return;
    }

    let timeoutId: NodeJS.Timeout | undefined;

    const handleWindowResize = () => {
      if (timeoutId) clearTimeout(timeoutId);

      if (delayTimeRef.current > 0) {
        timeoutId = setTimeout(updateWindowSize, delayTimeRef.current);
      } else {
        updateWindowSize();
      }
    };

    window.addEventListener("resize", handleWindowResize);
    return () => {
      window.removeEventListener("resize", handleWindowResize);
      if (timeoutId) clearTimeout(timeoutId);
    };
  }, [updateWindowSize]);
```
이 경우는 또 의존성 배열에 또 updateWindowSize추가를 해야 위에서 나타난 eslint오류가 발생하지 않습니다.
근데 제가 생각하기엔 저렇게 단순한 setState형태의 상태 업데이트를 분리해서 의존성까지 추가할 필요가 있나 싶어 의문이기도 하네요

아무튼 제가 생각한 최선은 updateWindowSize를 의존성배열에서 제거하고
`// eslint-disable-next-line react-hooks/exhaustive-deps`를 추가해서 eslint 규칙을 비활성화하는것이라고 생각하는데 어떤 방법이 가장 적절한 방법일까요?

Reply:
추가적으로 궁금한 점이 생겼습니다

좀 더 클린하고 정돈된 코드라 하면, 아래의 상황에서 

```ts
  useEffect(() => {
    if (!clientRef.current) {
      return;
    }

    let timeoutId: NodeJS.Timeout | null = null;

    const handleWindowResize = () => {
      if (timeoutId) clearTimeout(timeoutId);

      const updateWindowSize = () => {
        setWindowSize({
          width: window.innerWidth,
          height: window.innerHeight,
        });
      };

      if (delayTimeRef.current > 0) {
        timeoutId = setTimeout(updateWindowSize, delayTimeRef.current);
      } else {
        updateWindowSize();
      }
    };

    window.addEventListener("resize", handleWindowResize);

    return () => {
      window.removeEventListener("resize", handleWindowResize);
      if (timeoutId) clearTimeout(timeoutId);
    };
  }, []);
```
1. 상단의 현재상황 유지가 좋을지
2. useEffect안에 updateWindowSize를 밖으로 분리해서 useCallback을 사용하는 것이 더 적절한 방법인지 
3. 단순히 state를 업데이트하는 로직일 뿐이니 함수로 정의하지 않고 if문안에 setWindowSize를 두번 정의하는게 맞는지 궁금합니다


#### 2번
```ts
const updateWindowSize = useCallback(() => {
    setWindowSize({
      width: window.innerWidth,
      height: window.innerHeight,
    });
  }, []);

  useEffect(() => {
    if (!clientRef.current) {
      return;
    }

    let timeoutId: NodeJS.Timeout | null = null;

    const handleWindowResize = () => {
      if (timeoutId) clearTimeout(timeoutId);

      if (delayTimeRef.current > 0) {
        timeoutId = setTimeout(updateWindowSize, delayTimeRef.current);
      } else {
        updateWindowSize;
      }
    };

    window.addEventListener("resize", handleWindowResize);

    return () => {
      window.removeEventListener("resize", handleWindowResize);
      if (timeoutId) clearTimeout(timeoutId);
    };
  }, [updateWindowSize]);
```
이 경우, updateWindowSize함수는 한번 생성되고 재생성될 필요가 없는 함수이므 꼭 useEffect의 의존성 배열에 존재하지 않아도 된다고 생각해서 아래와 같은 방법이 더 적절할 수도 있다고 생각합니다
```ts
  // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);
```

#### 3번
```ts
 useEffect(() => {
    if (!clientRef.current) {
      return;
    }

    let timeoutId: NodeJS.Timeout | null = null;

    const handleWindowResize = () => {
      if (timeoutId) clearTimeout(timeoutId);

      if (delayTimeRef.current > 0) {
        timeoutId = setTimeout(
          () =>
            setWindowSize({
              width: window.innerWidth,
              height: window.innerHeight,
            }),
          delayTimeRef.current
        );
      } else {
        setWindowSize({
          width: window.innerWidth,
          height: window.innerHeight,
        });
      }
    };

    window.addEventListener("resize", handleWindowResize);

    return () => {
      window.removeEventListener("resize", handleWindowResize);
      if (timeoutId) clearTimeout(timeoutId);
    };
  }, []);
```

Reply:
> 추가적으로 궁금한 점이 생겼습니다 **updateWindowSize**(단순한 state 업데이트 함수)에 관련한 질문인데 좀 더 클린하고 정돈된 코드라 하면, 아래의 상황에서
> 
> #### 1. 상현재상황 유지
> useEffect내부에서 updateWindowSize를 정의함
> ...

저는 해당 함수가 다른 곳에서도 재활용 가능한 함수인지, 또는 가독성이 있는지를 고려하여 함수의 위치를 결정합니다. 현재 코드의 경우, useEffect 내부에서만 사용하는 함수라면, 저는 다음과 같이 작성할 것 같습니다.
```ts
...

useEffect(() => {
  let timeoutId: NodeJS.Timeout | undefined;

  const updateWindowSize = () => {
    setWindowSize({
      width: window.innerWidth,
      height: window.innerHeight,
    });
  };

  const handleWindowResize = () => {
    if (timeoutId) clearTimeout(timeoutId);

    if (isDelay) {
      timeoutId = setTimeout(updateWindowSize, delayTime);
    } else {
      updateWindowSize();
    }
  };

  window.addEventListener('resize', handleWindowResize);
  return () => {
    window.removeEventListener('resize', handleWindowResize);
    if (timeoutId) clearTimeout(timeoutId);
  };
}, [isDelay, delayTime]);

...
```

Reply:
@foresec 저는 2번 코드도 적절하다고 생각합니다!

말씀하신 바와 같이 재생성이 필요하지 않은 `updateWindowSize()` 함수는 밖으로 분리하고
갱신되는 `timeoutId`를 사용해야 하는 `handleWindowResize()` 함수는 useEffect 안에 작성하는 게 좋다고 생각합니다.

추가로 useCallback 훅을 작은 단위의 계산에서 사용하는 것은 지양하는 편인데요. 관련 글 링크를 첨부해 드릴테니 확인해보시면 좋을 것 같습니다.
- [useMemo, useCallback으로 최적화가 꼭 필요할까?](https://5kdk.github.io/blog/2023/07/04/necessary-with-usememo-and-usecallback)

---

추가로 🔗[useWindowSize](https://github.com/jbetancur/react-data-table-component/blob/59c6ac1b6a3004ac9e719815290524f6a5752bdc/src/hooks/useWindowSize.ts) 훅의 좋은 예시가 있어서 링크 첨부해드려요!

Reply:
@jeongbaebang @bicochan 고민 끝에 재사용성면에서 그 이상의 분리가 필요없다고 생각해 useEffect내부에서 분리하는 방법을 택했습니다. 하지만 종찬님께서 알려주신 의견과 정보글들 또한 앞으로의 react hook 활용성에 있어서 많은 도움이 될 것 같습니다!